src/core/ucc_coll.c

```diff
@@ -0,0 +1,40 @@
+/**
+ * Copyright (C) Mellanox Technologies Ltd. 2021.  ALL RIGHTS RESERVED.
+ * See file LICENSE for terms.
+ */
+
+#include "config.h"
+#include "ucc_team.h"
+#include "ucc_context.h"
+#include "components/cl/ucc_cl.h"
+#include "utils/ucc_malloc.h"
+#include "utils/ucc_log.h"
+
+static ucc_cl_team_t *ucc_select_cl_team(ucc_coll_op_args_t *coll_args,
+                                         ucc_team_t *team)
+{
+    /* TODO: collective CL selection logic will be there.
+       for now just return 1st CL on a list */
+    return team->cl_teams[0];
+}
+
+ucc_status_t ucc_collective_init(ucc_coll_op_args_t *coll_args,
+                                 ucc_coll_req_h *request, ucc_team_h team)
+{
+    ucc_cl_team_t          *cl_team;
+    ucc_base_coll_op_args_t op_args;
+    ucc_status_t            status;
+    ucc_coll_task_t        *task;
+    /* TO discuss: maybe we want to pass around user pointer ? */
+    memcpy(&op_args.args, coll_args, sizeof(ucc_coll_op_args_t));
```

Comment:
Making a copy might be expensive for some types of collectives that have data buffers proportional to the number of ranks/pes in the collectives. Good to have a discussion on this.

Reply:
@manjugv So, this memcpy is actually just copying our coll_op_args (which is 120 bytes at the moment) regardless of the scale or coll type. However, it works only for the "non-vector" collectives, ie those that don't have multiple src,dest buffers or offset, dtypes arrays. In the latter case we will need a better solution. 

1st option that always works - deep copying the input - might be very expensive (though collectives like Alltoallv, allgatherv are very expensive on their own).
2nd option - we can have a couple of flags on the ucc_coll_op_args_t (we can use "mask" to store them, or actually add "flags" field). 1 such flag would be a hint, UCC_COLL_ARGS_PERSISTENT - this would tell the library that USER will guarantee that coll_op_args structure is not released or altered until the collective request is finalized. Another useful flag i have in mind is UCC_COLL_BLOCKING. This flag would tell the library that user invoking the collective will do the loop of progress/test right after the coll start and will not post any other collectives untils this one completes. This will be used for blocking MPI colls implementation and internally we will use higher polling settings in the TLs for this mode.